### PR TITLE
DEP 4: Release Schedule

### PR DESCRIPTION
**Pull request by timgraham**

### CODE REVIEW

**Comment by timgraham** [comment]
I'm not sure what the exact scope of this DEP should be (e.g. if things like Python support and the FAQ should be included.)

File: final/0002-release-schedule.rst [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+see at a glance how compatible releases are with each other. It also helps to
+anticipate when compatibility shims will be removed. It's not a pure form of
+SemVer as each 8 month feature release will continue to have some documented
+backwards incompatibilities where a deprecation path isn't possible or not
+worth the cost. Also deprecations from the last minor release in a series will
+be dropped in a non-dot-zero release to accommodate our policy of keeping
+deprecation shims for at least two releases.
+
+Starting with Django 3.0, each version following an LTS will bump to the next
+"dot zero" version. LTS version will thereafter always be "X.2". The
```
Comment (MarkusH) [inline final/0002-release-schedule.rst]: LTS version**s**

File: final/0002-release-schedule.rst [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+deprecation policy outlined below will make it feasible for apps to support
+two LTS versions, and so third-party apps can target support for X.2 - X+1.2.
```
Comment (MarkusH) [inline final/0002-release-schedule.rst]: support from X.2 to X+1.2 ?

File: final/0002-release-schedule.rst [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+Deprecation Policy
+==================
+
+The deprecation policy is designed to ease upgrades from one LTS to the next
```
Comment (MarkusH) [inline final/0002-release-schedule.rst]: LTS *version*

File: final/0002-release-schedule.rst [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+support of two consecutive LTS releases.
+
+Here's an overview of when deprecated features will be dropped in Django:
+
+* X.2: (LTS) ---
+* Y.0: ---
+* Y.1: Drop deprecation shims added in Django X.2 (older version are no longer
+  supported).
+* Y.2 (LTS): No deprecation shims dropped (Y.0 is no longer supported, but
+  keeping the shims allows supporting easily supporting X.2 where the
```
Comment (MarkusH) [inline final/0002-release-schedule.rst]: "keeping the shims allows for easy support of X.2 where"?

File: final/0002-release-schedule.rst [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+  replacements for features deprecated in Y.0 aren't present).
+* Z.0: Drop deprecation shims added in Y.1.
+
+Python support
+==============
+
+We will support a Python version up to and including the first Django LTS
```
Comment (MarkusH) [inline final/0002-release-schedule.rst]: Mind rephrasing that sentence. It's hard to parse and understand.

File: final/0002-release-schedule.rst [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+September 2017 and Django 1.8 LTS security support ends April 2018. Therefore
+Django 1.8 is the last version to support Python 3.3.
+
+Frequently Asked Questions and Complaints
+=========================================
+
+**Q: Will a time-based release compromise the stability of releases? Why not
+release when things are ready?**
+
+A: Release stability will be a continued priority and releases will be delayed
```
Comment (MarkusH) [inline final/0002-release-schedule.rst]: This does not answer "Why not release when things are ready?" in my eyes.

Comment (timgraham) [inline final/0002-release-schedule.rst]: The main point is "releases will be delayed as necessary if there are major, unsolved issues." Can you rephrase what question remains unanswered?

File: final/0002-release-schedule.rst (hunk 5 above)
Comment (timgraham) [inline final/0002-release-schedule.rst]: We will support a Python version up to and including the first
Django LTS release whose security support ends after security support for that
version of Python ends.

Comment (MarkusH) [inline final/0002-release-schedule.rst]: much better. Thanks

File: final/0002-release-schedule.rst (hunk 6 above)
Comment (MarkusH) [inline final/0002-release-schedule.rst]: On second read, the question is indeed answered. Maybe just split the first sentence in two to clarify things?

"Release stability will continue to be our main priority. If there are major, unsolved issues, we will delay a release as necessary."?

File: final/0002-release-schedule.rst [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+
+The deprecation policy is designed to ease upgrades from one LTS release to the
+next and to allow third-party apps easily support the currently supported
+Django versions. The policy should make it easy to develop an app using the
+latest LTS release and have it continue to work without many changes until the
+next LTS. Upgrading from LTS to LTS should be easy enough if you don't use any
+deprecated features when running on the older LTS.
+
+Features will be deprecated over two releases, except for features deprecated
+in the version after an LTS as those deprecation shims are needed to ease
```
Comment (carljm) [inline final/0002-release-schedule.rst]: add "which will be deprecated over three releases" - right now it's unclear what the exception actually is.

File: final/0002-release-schedule.rst [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+Here's an overview of when deprecated features will be dropped in Django:
+
+* X.2: (LTS) ---
+* Y.0: ---
+* Y.1: Drop deprecation shims added in Django X.2 (older version are no longer
+  supported).
+* Y.2 (LTS): No deprecation shims dropped (Y.0 is no longer supported, but
+  keeping the shims allows easy support of X.2 where the replacements for
+  features deprecated in Y.0 aren't present).
+* Z.0: Drop deprecation shims added in Y.1.
```
Comment (carljm) [inline final/0002-release-schedule.rst]: This should be "drop deprecation shims added in Y.0 and Y.1."

**Comment by carljm** [comment]
I think this looks pretty good! I would actually add a short "Immediate Future" section towards the end, clarifying how the transition to the new numbering scheme will occur (that is, the 1.9 - 3.0 schedule).

**Comment by MarkusH** [comment]
There is already a DEP 0002 in "draft" and 0003 in "accepted"

**Comment by timgraham** [comment]
Carl, my thought is to communicate "Immediate Future" in the blog post and download page roadmap. I don't think it needs to be memorialized in a DEP.

**Comment by carljm** [comment]
If the model here is the PEP process, PEPs do usually contain necessary information on the transition, too, even if that information won't be relevant years later. It's not the focus, but it's part of the PEP.

I guess I don't see the benefit of leaving it out entirely; if someone is reading the DEP in the near term, how the transition will happen is very useful information. Since it's even less likely that we want to link from a DEP to a more-transient resource like a blog post or the download page, not providing the information just leaves a reader in the dark. And it seems like the information would only take a sentence or two to provide (especially with the new 1.10 - 1.11 plan).

**Comment by timgraham** [comment]
I did add "effective with Django 2.0 which will follow Django 1.11 (LTS)."

**Comment by carljm** [comment]
Ah, missed that change! DEP LGTM; are you ready for a technical board vote on it?

**Comment by timgraham** [comment]
Yes, I think it's ready now.
